subsys/mgmt/ec_host_cmd/backends/ec_host_cmd_backend_uart.c

```diff
@@ -0,0 +1,293 @@
+	case USART_HOST_CMD_RX_OVERRUN:
+		/* If state is rx_overrun then timeout was hit because
+		 * process request was cancelled and extra rx bytes were
+		 * dropped
+		 */
+		LOG_ERR("Request overrun detected");
+		break;
+	case USART_HOST_CMD_RX_BAD:
+		/* If state is rx_bad then packet header was bad and process
+		 * request was cancelled to drop all incoming bytes.
+		 */
+		LOG_ERR("Bad packet header detected");
+		break;
+	default:
+		LOG_ERR("Request timeout mishandled, state: %d", hc_uart->state);
+	}
+
+	/* TODO: Workaround - reprogram DMA not to lose data at the end of buffer.
+	 * Remove it once the circular buffer is supported.
+	 */
+	res = uart_rx_disable(hc_uart->uart_dev);
+	res = uart_rx_enable(hc_uart->uart_dev, hc_uart->rx_ctx->buf,
+			     CONFIG_EC_HOST_CMD_HANDLER_RX_BUFFER, UART_OVERRUN_TIMEOUT_US);
+
+	hc_uart->state = USART_HOST_CMD_READY_TO_RX;
+}
+
+static void uart_callback(const struct device *dev, struct uart_event *evt, void *user_data)
+{
+	struct ec_host_cmd_uart_ctx *hc_uart = user_data;
+
+	switch (evt->type) {
+	case UART_RX_RDY:
+		uint8_t *evt_rx_buff = evt->data.rx.buf + evt->data.rx.offset;
+		uint8_t *rx_ctx_buff = hc_uart->rx_ctx->buf + hc_uart->rx_ctx->len;
+		size_t new_len = hc_uart->rx_ctx->len + evt->data.rx.len;
+
+		if (hc_uart->state == USART_HOST_CMD_READY_TO_RX) {
+			hc_uart->rx_ctx->len = 0;
+			hc_uart->state = USART_HOST_CMD_RECEIVING;
+			k_work_reschedule(&hc_uart->timeout_work, UART_REQ_RX_TIMEOUT);
+		} else if (hc_uart->state == USART_HOST_CMD_PROCESSING) {
+			LOG_ERR("USART HOST CMD ERROR: Received data while processing a command");
+			return;
+		}
+
+		/* Copy data only if we are in the receiving state.*/
+		if ((hc_uart->state == USART_HOST_CMD_RECEIVING) &&
+		    (new_len <= CONFIG_EC_HOST_CMD_HANDLER_RX_BUFFER)) {
+			memcpy(rx_ctx_buff, evt_rx_buff, evt->data.rx.len);
+		} else if (new_len > CONFIG_EC_HOST_CMD_HANDLER_RX_BUFFER) {
+			hc_uart->state = USART_HOST_CMD_RX_BAD;
+			return;
+		}
+
+		hc_uart->rx_ctx->len += evt->data.rx.len;
+
+		if (hc_uart->rx_ctx->len >= sizeof(struct ec_host_cmd_request_header)) {
+			/* Buffer has request header. Check header and get data_len */
+			hc_uart->expected_len = request_expected_size(
+				(struct ec_host_cmd_request_header *)hc_uart->rx_ctx->buf);
+
+			if (hc_uart->expected_len == 0 ||
+			    hc_uart->expected_len > CONFIG_EC_HOST_CMD_HANDLER_RX_BUFFER) {
+				/* Invalid expected size, set the state and wait for timeout */
+				hc_uart->state = USART_HOST_CMD_RX_BAD;
+			} else if (hc_uart->rx_ctx->len == hc_uart->expected_len) {
+				/* Don't wait for overrun, because it is already done
+				 * in a UART driver.
+				 */
+				(void)k_work_cancel_delayable(&hc_uart->timeout_work);
+
+				/* TODO: Workaround - reprogram DMA not to lose data at the end of
+				 * buffer. Remove it once the circular buffer is supported.
+				 */
+				uart_rx_disable(hc_uart->uart_dev);
+				uart_rx_enable(hc_uart->uart_dev, hc_uart->rx_ctx->buf,
```

Comment:
We are passing `hc_uart->rx_ctx->buf` buffer to UART driver, but we are also copying data to that buffer in line 164. Looks like the data can be overwritten here when processing the host command.

Reply:
True. Fixed.